docs/develop-an-agent.md

```diff
@@ -0,0 +1,423 @@
+```
+
+More details on how the IoT Hub service interprets this signed message can be found in the docs on [Security tokens](https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-security#security-tokens). 
+
+
+#### **Step 3** Get a signed token from the keys service
+
+Send a request to the key service endpoint `/run/aziot/keyd.sock`.
+
+If the connection is refused, go back to Step 1 since IoT Hub device information could have changed.
+
+Otherwise, parse the `signature` from the resposne.
+
+```bash
+# Request the key service sign the message and then parse the 'signature' from the response and url decode it
+signature="$(
+    sudo curl --unix-socket '/run/aziot/keyd.sock' \
+        -X POST -H 'content-type: application/json' --data-raw "$(
+            jq -cn --arg 'keyHandle' "$key_handle" --arg 'message' "$signature_data" '{
+                "keyHandle": $keyHandle,
+                "algorithm": "HMAC-SHA256",
+                "parameters": {
+                    "message": $message,
+                },
+            }'
+        )" 'http://keyd.sock/sign?api-version=2020-09-01' |
+        jq '.signature' -r |
+        sed -e 's|+|%2b|g' -e 's|/|%2f|g' -e 's|=|%3d|g'
+)"
+sas_token="$(printf 'sr=%s&se=%s&sig=%s' "$resource_uri" "$expiry" "$signature")"
+```
+
+#### **Step 4**: Create the full connection string
+
+If the `gatewayHost` retrieved from the identity information is the same as the `hubName`, then the device is configured to connect directly to IoT Hub. Alternatively, if they differ then the device is configured to be behind a gateway.
+
+```bash
+# Note that the 'SharedAccessSignature' appears both in the connection string and in the value for it.
+sas="SharedAccessSignature $sas_token"
+
+if [ $hub_name = $gateway_host ]; then
+    # Connecting directly to Hub.
+    echo "HostName=$hub_name;DeviceId=$device_id;ModuleId=$module_id;SharedAccessSignature=$sas"
+else
+    # In a nested configuration. Include the gateway name in the connection string.
+    echo "HostName=$hub_name;DeviceId=$device_id;ModuleId=$module_id;SharedAccessSignature=$sas;GatewayHost=$gateway_host"
+fi
+
+```
+
+> **Note**
+>
+> Portions of the script above come from the [iothub-get-twin.sh](https://github.com/Azure/iot-identity-service/blob/main/ci/iothub-get-twin.sh) script used in the tests for the identity service.
+
+
+#### **Step 5** Connect to IoT Hub using the SDK
+
+Connect to IoT Hub using the C SDK's [IoTHubDeviceClient_LL_CreateFromConnectionString](https://docs.microsoft.com/en-us/azure/iot-hub/iot-c-sdk-ref/iothub-device-client-ll-h/iothubdeviceclient-ll-createfromconnectionstring) API.
+
+#### **Step 6** After disconnect,
+
+Repeat all steps starting from _Step 1_ above.
+
+## Using the C SDK to connect to IoT Hub for SaS authentication
+
+The steps below are how to use the C SDK in your agent to establish IoT Hub connectivity using a module identity and using the SharedAccessSignature connection string method.
```

Comment:
```suggestion
The steps below use the C SDK to establish IoT Hub connectivity with a module identity and the SharedAccessSignature connection string method.
```

Reply:
I'm going to remove this section since it essentially repeats the prior steps 1 - 6.